src/world/area_kgr/kgr_01/8A9B70.c

```diff
@@ -1,9 +1,98 @@
 #include "kgr_01.h"
 
-INCLUDE_ASM(s32, "world/area_kgr/kgr_01/8A9B70", func_80240160_8A9B70);
+void mdl_get_copied_vertices(s32 copyIndex, Vtx** firstVertex, Vtx** copiedVertices, s32* numCopied);
+Gfx* mdl_get_copied_gfx(s32 copyIndex);
 
-INCLUDE_ASM(s32, "world/area_kgr/kgr_01/8A9B70", func_80240354_8A9D64);
+#define TONGUE_WIGGLE_RATE 10
 
-INCLUDE_ASM(s32, "world/area_kgr/kgr_01/8A9B70", func_802403E0_8A9DF0);
+extern u16 N(TongueWiggleTime); // tongue wiggle rate
 
-INCLUDE_ASM(s32, "world/area_kgr/kgr_01/8A9B70", func_80240500_8A9F10);
+void N(add_tongue_deformation)(Vtx* src, Vtx* dest, s32 numVertices, s32 time) {
+    PlayerStatus *player = &gPlayerStatus;
+    Vtx_t* vs;
+    Vtx_t* vd;
+    s32 dx;
+    s32 dz;
+    s32 offset;
+    s32 pct;
+    s32 i;
+    
+    for (i = 0; i < numVertices; i++) {
+        vs = &src[i].v;
+        vd = &dest[i].v;
+
+        // base y-offset goes from 0-5 based on radial distance to player,
+        // with 0 at the closest and 5 when distance squared > 1000.
+        // this creates the depression of the tongue where the player is standing.
+        dx = vd->ob[0] - player->position.x;
+        dz = vd->ob[2] - player->position.z;
+        offset = ((dx * dx) + (dz * dz)) / 100;
+        if (offset > 10) {
+          offset = 10;
+        }
+        offset /= 2;
+
+        // space-based and time-based undulations are the second componet of the y-offset
+        // reuse of the pct variable is not ideal, but required to match
+        pct = 2.0f + 2.0f * sin_rad(
+            (f32)vd->ob[0] * 0.06 +
+            (f32)vd->ob[2] * 0.03f +
+            time * 0.01f);
+        offset += pct;
+
+        // modulate y-offset based on vertex x-position
+        // this anchors the tongue in the back of the throat and makes the unudulations
+        // more prominent closer toward the tip of the tongue
+        pct = 200 - vd->ob[0];
+        if (pct < 0) {
+          pct = 0;
+        }
+        if (pct > 100) {
+          pct = 100;
+        }
+        offset = offset * pct / 100;
+
+        vd->ob[1] = vs->ob[1] + offset;
+    }
+}
+
+// conforms to ModelCustomGfxBuilderFunc
+void N(make_tongue_gfx)(int index) {
+    Vtx* vtxSrc;
```

Comment:
Don't these wanna be non-pointers technically? So there's actual space for them allocated on the stack properly? It may match by default but yeah. idk

Reply:
Also the `int index` should probably be `s32 index`

Reply:
no, we're manipulating elements of the dest vertex array; we dont want a local copy. the vs/vd are needed to match.